group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2684,8 +2804,160 @@ static Errors appendGroupMetadataErrorToResponseError(Errors appendError) {
+                group,
+                request.memberId(),
+                request.groupInstanceId(),
+                "sync-group"
+            );
+            if (memberError.isPresent()) {
+                return memberError;
+            } else {
+                if (request.generationId() != group.generationId()) {
+                    return Optional.of(Errors.ILLEGAL_GENERATION);
+
+                } else if (isProtocolInconsistent(request.protocolType(), group.protocolType().orElse(null)) ||
+                    isProtocolInconsistent(request.protocolName(), group.protocolName().orElse(null))) {
+
+                    return Optional.of(Errors.INCONSISTENT_GROUP_PROTOCOL);
+                } else {
+                    return Optional.empty();
+                }
+            }
+        }
+    }
+
+    private void removePendingSyncMember(
+        GenericGroup group,
+        String memberId
+    ) {
+        group.removePendingSyncMember(memberId);
+
+        String syncKey = genericGroupSyncKey(group.groupId());
+
+        if (group.generationId() != group.pendingSyncGenerationId()) {
+            timer.cancel(syncKey);
+        } else {
+            switch (group.currentState()) {
+                case DEAD:
+                case EMPTY:
+                case PREPARING_REBALANCE:
+                    timer.cancel(syncKey);
+                    break;
+                case COMPLETING_REBALANCE:
+                case STABLE:
+                    if (group.hasReceivedSyncFromAllMembers()) {
+                        timer.cancel(syncKey);
+                    }
+                    break;
+                default:
+                    throw new IllegalStateException("Unknown group state: " + group.stateAsString());
+            }
+        }
+    }
+
+    /**
+     * Handle a generic group HeartbeatRequest.
+     *
+     * @param context        The request context.
+     * @param request        The actual Heartbeat request.
+     *
+     * @return The Heartbeat response.
+     */
+    public HeartbeatResponseData genericGroupHeartbeat(
+        RequestContext context,
+        HeartbeatRequestData request
+    ) {
+        GenericGroup group = getOrMaybeCreateGenericGroup(request.groupId(), false);
+
+        validateGenericGroupHeartbeat(group, request.memberId(), request.groupInstanceId(), request.generationId());
+
+        switch (group.currentState()) {
+            case EMPTY:
+                return new HeartbeatResponseData().setErrorCode(Errors.UNKNOWN_MEMBER_ID.code());
+
+            case PREPARING_REBALANCE:
+                rescheduleGenericGroupMemberHeartbeat(group, group.member(request.memberId()));
+                return new HeartbeatResponseData().setErrorCode(Errors.REBALANCE_IN_PROGRESS.code());
+
+            case COMPLETING_REBALANCE:
+                // Consumers may start sending heartbeat after join-group response, in which case
+                // we should treat them as normal hb request and reset the timer
+
```

Comment:
This one reads a bit weird. It may be better to put it together with STABLE and to rework the comment to say that it applies to the completing rebalance.

```
case COMPLETING_REBALANCE:
case STABLE:
// Comment...
```